eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/AppendCondition.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2010-2024. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventsourcing.eventstore;
+
+import jakarta.annotation.Nonnull;
+import org.axonframework.eventhandling.EventMessage;
+
+/**
+ * Interface describing the consistency boundary condition for
+ * {@link org.axonframework.eventhandling.EventMessage EventMessages} when
+ * {@link EventStoreTransaction#appendEvent(EventMessage) appending} them to an Event Store.
+ *
+ * @author Michal Negacz
+ * @author Milan Savić
+ * @author Marco Amann
+ * @author Sara Pellegrini
+ * @author Steven van Beelen
+ * @since 5.0.0
+ */
+public sealed interface AppendCondition permits NoAppendCondition, DefaultAppendCondition {
```

Comment:
Fair question! The intent was to ensure on the interface level that **nobody** can provide an `AppendCondition` where there's a consistency marker without any criteria, and vice versa. 

If the interface could be implemented by others, the freedom would exists to wrongfully call an Event Store.

Granted, we could put the check in the Event Store, but I figured this might be a fair angle to pursue instead.
This was very much a "ow maybe this is nice!" idea, so I am completely fine if we just add a basic `AppendCondition` validation to every Store eventually.